ecs-agent/manageddaemon/managed_daemon.go

```diff
@@ -0,0 +1,326 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//      http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package manageddaemon
+
+import (
+	"fmt"
+	"time"
+
+	dockercontainer "github.com/docker/docker/api/types/container"
+)
+
+const (
+	imageTarPath                          = "/var/lib/ecs/deps/daemons"
+	imageTagDefault                       = "latest"
+	defaultAgentCommunicationPathHostRoot = "/var/run/ecs"
+	defaultApplicationLogPathHostRoot     = "/var/log/ecs"
+	defaultAgentCommunicationMount        = "agentCommunicationMount"
+	defaultApplicationLogMount            = "applicationLogMount"
+)
+
+type ManagedDaemon struct {
+	imageName string
+	imageRef  string
+	imageTag  string
+
+	healthCheckTest     []string
+	healthCheckInterval time.Duration
+	healthCheckTimeout  time.Duration
+	healthCheckRetries  int
+
+	// Daemons require an agent <-> daemon mount
+	// identified by the volume name `agentCommunicationMount`
+	// the SourceVolumeHostPath will always be overridden to
+	// /var/run/ecs/<md.imageName>
+	agentCommunicationMount *MountPoint
+
+	// Daemons require an application log mount
+	// identified by the volume name `applicationLogMount`
+	// the SourceVolumeHostPath will always be overridden to
+	// /var/log/ecs/<md.imageName>
+	applicationLogMount *MountPoint
+
+	mountPoints []*MountPoint
+	environment map[string]string
+
+	loadedDaemonImageRef string
+}
+
+// A valid managed daemon will require
+// healthcheck and mount points to be added
+func NewManagedDaemon(
+	imageName string,
+	imageRef string,
+	imageTag string,
+) *ManagedDaemon {
+	if imageTag == "" {
+		imageTag = imageTagDefault
+	}
+	// health check retries 0 is valid
```

Comment:
did you mean health check retries 0 is *invalid*?

Reply:
https://docs.docker.com/engine/reference/builder/ 
I see `--retries=N (default: 3)` but see varying thoughts about whether 0 is valid.  As we set this to -1 by default, this validation just checks that we've set this to a non-default number.